### PR TITLE
Feature upload notifications for concent

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/667

### CODE REVIEW

File: concent_api/conductor/tests/test_integration_conductor.py [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 from core.tests.utils import ConcentIntegrationTestCase
-from ..models       import BlenderSubtaskDefinition
-from ..models       import UploadReport
-from ..models       import VerificationRequest
-from ..tasks        import blender_verification_request
+from ..models import BlenderSubtaskDefinition
+from ..models import ResultTransferRequest
+from ..models import UploadReport
+from ..models import VerificationRequest
+from ..tasks import blender_verification_request
```
Comment (Jakub89) [inline concent_api/conductor/tests/test_integration_conductor.py]: Can't we use `from core.tasks ...` and `from core.models ...`? It would be more readable I think.

File: concent_api/conductor/models.py [hunk 2]
```diff
@@ -90,6 +107,9 @@ class UploadReport(Model):
     # Foreign key to VerificationRequest. Can be NULL if there's no corresponding request.
     verification_request = ForeignKey(VerificationRequest, related_name='upload_reports', blank=True, null=True)
 
+    # Foreign key to z. Can be NULL if there's no corresponding request.
```
Comment (Jakub89) [inline concent_api/conductor/models.py]: Foreign key to what?

File: concent_api/conductor/models.py [hunk 3]
```diff
@@ -90,6 +107,9 @@ class UploadReport(Model):
...
+    result_transfer_request = ForeignKey(ResultTransferRequest, related_name='upload_reports', blank=True, null=True)
+
     # Indicates when conductor has been notified about the upload.
     created_at      = DateTimeField(default=timezone.now)
```
Comment (Jakub89) [inline concent_api/conductor/models.py]: You can fix style in here I think. It's the last place in this file with our old one.

File: concent_api/core/tests/test_auth_get_task_result.py [hunk 4]
```diff
@@ -7,25 +7,18 @@
...
+from conductor.service import result_upload_finished
+from core.models import PendingResponse
+from core.models import Subtask
+from core.tests.utils import ConcentIntegrationTestCase
 from core.tests.utils import parse_iso_date_to_timestamp
-from common.constants        import ErrorCode
-from common.testing_helpers  import generate_ecc_key_pair
 
 
 (CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY)                         = generate_ecc_key_pair()
```
Comment (kbeker) [inline concent_api/core/tests/test_auth_get_task_result.py]: Please change style here to our actual convention

File: concent_api/core/tests/test_auth_get_task_result.py [hunk 5]
```diff
@@ -70,12 +63,15 @@ def test_requestor_forces_get_task_result_and_concent_immediately_sends_rejectio
...
-                reverse('core:send'),
-                data                                = original_serialized_force_get_task_result,
-                content_type                        = 'application/octet-stream',
-            )
+        with mock.patch('conductor.tasks.result_transfer_request.delay') as result_transfer_request_task:
+            with freeze_time("2017-12-01 11:00:08"):
+                response = self.client.post(
+                    reverse('core:send'),
+                    data                                = original_serialized_force_get_task_result,
+                    content_type                        = 'application/octet-stream',
```
Comment (kbeker) [inline concent_api/core/tests/test_auth_get_task_result.py]: Please change style here

File: concent_api/core/tests/test_auth_get_task_result.py [hunk 6]
```diff
@@ -488,15 +474,17 @@ def test_concent_requests_task_result_from_provider_and_requestor_receives_failu
...
+                response = self.client.post(
+                    reverse('core:send'),
+                    data=serialized_force_get_task_result,
+                    content_type='application/octet-stream',
+                )
 
-        self.assertEqual(response.status_code,      200)
+        result_transfer_request_task.assert_called_once()
 
+        self.assertEqual(response.status_code,      200)
```
Comment (kbeker) [inline concent_api/core/tests/test_auth_get_task_result.py]: Style please

File: concent_api/core/tests/test_integration_get_task_result.py [hunk 7]
```diff
@@ -66,12 +59,15 @@ def test_requestor_forces_get_task_result_and_concent_immediately_sends_rejectio
...
-                data                                = serialized_force_get_task_result,
-                content_type                        = 'application/octet-stream',
-            )
+        with mock.patch('conductor.tasks.result_transfer_request.delay') as result_transfer_request_task:
+            with freeze_time("2017-12-01 11:02:31"):
+                response = self.client.post(
+                    reverse('core:send'),
+                    data                                = serialized_force_get_task_result,
+                    content_type                        = 'application/octet-stream',
+                )
```
Comment (kbeker) [inline concent_api/core/tests/test_integration_get_task_result.py]: Style please

File: concent_api/core/tests/test_integration_get_task_result.py [hunk 8]
```diff
@@ -190,12 +192,15 @@ def test_requestor_forces_get_task_result_and_concent_immediately_sends_rejectio
...
-                reverse('core:send'),
-                data                                = serialized_force_get_task_result,
-                content_type                        = 'application/octet-stream',
-            )
+        with mock.patch('conductor.tasks.result_transfer_request.delay') as result_transfer_request_task:
+            with freeze_time("2017-12-01 11:00:08"):
+                response = self.client.post(
+                    reverse('core:send'),
+                    data                                = serialized_force_get_task_result,
+                    content_type                        = 'application/octet-stream',
```
Comment (kbeker) [inline concent_api/core/tests/test_integration_get_task_result.py]: Style please

File: concent_api/core/tests/test_integration_get_task_result.py [hunk 9]
```diff
@@ -1026,20 +1020,14 @@ def test_concent_requests_task_result_from_provider_and_requestor_receives_task_
...
-        )
+        with freeze_time("2017-12-01 11:00:54"):
+            response_3 = self.client.post(
+                reverse('core:receive'),
+                data=self._create_requestor_auth_message(),
+                content_type='application/octet-stream',
+            )
 
         self.assertEqual(response_3.status_code,  200)
         self._test_subtask_state(
```
Comment (kbeker) [inline concent_api/core/tests/test_integration_get_task_result.py]: Good thing might be to change style in whole file if you touched many places there. Please do it as a separate commit

File: concent_api/core/tests/test_task_result_upload_finished.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+class CoreResultUploadFinishedTaskTransactionTest(TransactionTestCase):
+
+    multi_db = True
+
+    def setUp(self):
+        super().setUp()
+
+        (self.PROVIDER_PRIVATE_KEY, self.PROVIDER_PUBLIC_KEY) = generate_ecc_key_pair()
+        (self.REQUESTOR_PRIVATE_KEY, self.REQUESTOR_PUBLIC_KEY) = generate_ecc_key_pair()
```
Comment (kbeker) [inline concent_api/core/tests/test_task_result_upload_finished.py]: Please move it `TransactionTestCase`

File: concent_api/conductor/tests/test_integration_conductor.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/conductor/tests/test_integration_conductor.py]: :+1: done

File: concent_api/conductor/models.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/conductor/models.py]: :+1: good catch, done

File: concent_api/conductor/models.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/conductor/models.py]: :+1: done

File: concent_api/core/tests/test_auth_get_task_result.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_auth_get_task_result.py]: :+1: done

File: concent_api/core/tests/test_auth_get_task_result.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_auth_get_task_result.py]: :+1: done

File: concent_api/core/tests/test_auth_get_task_result.py (hunk 6 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_auth_get_task_result.py]: :+1: done

File: concent_api/core/tests/test_integration_get_task_result.py (hunk 7 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_integration_get_task_result.py]: :+1: done

File: concent_api/core/tests/test_integration_get_task_result.py (hunk 8 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_integration_get_task_result.py]: :+1: done

File: concent_api/core/tests/test_integration_get_task_result.py (hunk 9 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_integration_get_task_result.py]: :+1: done

File: concent_api/core/tests/test_task_result_upload_finished.py (hunk 10 above)
Comment (rwrzesien) [reply concent_api/core/tests/test_task_result_upload_finished.py]: `TransactionTestCase` is taken directly from Django, I can't modify it.

File: concent_api/conductor/service.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+from conductor.models import ResultTransferRequest
+from conductor.models import UploadReport
+from conductor.models import VerificationRequest
+from core.tasks import result_upload_finished
+
+
+logger = logging.getLogger(__name__)
+
+
+def report_finished_upload(file_path: str, result_transfer_request: ResultTransferRequest) -> None:
```
Comment (dybi) [inline concent_api/conductor/service.py]: can this be called with `result_transfer_request` where `result_transfer_request.upload_finished = True`?

File: concent_api/conductor/tests/test_service.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                self.path,
+                self.result_transfer_request,
+            )
+
+        upload_report.refresh_from_db()
+        self.assertEqual(upload_report.result_transfer_request, self.result_transfer_request)
+
+        result_upload_finished.assert_not_called()
+
+    def test_that_report_finished_upload_should_result_upload_finished_if_result_transfer_request_upload_finished_is_false(self):
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/conductor/tests/test_service.py]: :+1: done

File: concent_api/conductor/models.py [hunk 13]
```diff
@@ -53,6 +53,23 @@ def clean(self) -> None:
...
 
+class ResultTransferRequest(Model):
+    """
+    Transfer request created during `get-task-results` use case.
+    """
+
+    subtask_id = CharField(max_length=MESSAGE_TASK_ID_MAX_LENGTH)  # TODO: Unique?
+
+    # Relative path of the .zip file that contains the rendering result received from the provider.
+    result_package_path = TextField(validators=[MaxLengthValidator(MESSAGE_PATH_LENGTH)])  # TODO: Unique?
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/conductor/models.py]: I think both of them - `path` and `subtask_id`. @cameel Can you confirm? If yes, maybe issue should be updated?

File: concent_api/conductor/service.py (hunk 11 above)
Comment (rwrzesien) [reply concent_api/conductor/service.py]: Yes, this object may exists after `get-task-results` use case was finished, and we are in `verification` use case, and Conductor's `report_upload` endpoint is called.

File: concent_api/core/tests/test_task_result_upload_finished.py (hunk 10 above)
Comment (kbeker) [reply concent_api/core/tests/test_task_result_upload_finished.py]: O, I didn't know I need to read about it ;)

Comment (rwrzesien) [reply concent_api/core/tests/test_task_result_upload_finished.py]: Check the import :)

File: concent_api/conductor/service.py (hunk 11 above)
Comment (dybi) [inline concent_api/conductor/service.py]: I have been thinking about it for a while and the name doesn't seem right. 'finished_upload' is reported conditionally. What always happens in this function is updating `UploadReport`. Thus maybe this function should be renamed to `update_upload_report`? I dnon't know. something bothers me here, but I am not sure what

File: concent_api/conductor/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (cameel) [reply concent_api/conductor/models.py]: Yes, they should be unique because `subtask_id` is unique and we cannot have more than one result package in a single subtask.

File: concent_api/conductor/models.py [hunk 14]
```diff
@@ -53,6 +53,23 @@ def clean(self) -> None:
             })
 
 
+class ResultTransferRequest(Model):
+    """
+    Transfer request created during `get-task-results` use case.
+    """
+
+    subtask_id = CharField(max_length=MESSAGE_TASK_ID_MAX_LENGTH)  # TODO: Unique?
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/conductor/models.py]: :+1: done

File: concent_api/conductor/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/conductor/models.py]: :+1: done
